src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.code/src/jdk/vm/ci/code/BytecodeFrame.java

```diff
@@ -351,30 +351,38 @@ public BytecodeFrame caller() {
     }
 
     @Override
-    public int hashCode() {
-        return (numLocals + 1) ^ (numStack + 11) ^ (numLocks + 7);
+    public int hashCode()
+        return Objects.hash(super.hashCode(),
+                            duringCall,
+                            numLocals,
+                            numLocks,
+                            numStack,
+                            rethrowException,
+                            Arrays.hashCode(slotKinds),
+                            Arrays.hashCode(values));
     }
 
     @Override
     public boolean equals(Object obj) {
         if (this == obj) {
             return true;
         }
-        if (obj instanceof BytecodeFrame && super.equals(obj)) {
-            BytecodeFrame that = (BytecodeFrame) obj;
-            // @formatter:off
-            if (this.duringCall == that.duringCall &&
-                this.rethrowException == that.rethrowException &&
-                this.numLocals == that.numLocals &&
-                this.numLocks == that.numLocks &&
-                this.numStack == that.numStack &&
-                Arrays.equals(this.values, that.values)) {
-                return true;
-            }
-            // @formatter:off
-            return true;
+        if (!super.equals(obj)) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
```

Comment:
Is this comparison better/faster than instanceof, and allows the compiler to avoid repeating the same check on the checkcast below?

Reply:
It's what Eclipse auto-generates for `equals`. I can confirm that Graal eliminates the check cast but don't think it's worth worrying about if not all compilers will.